### ISSUE TITLE
FR: real time progress in Gcode viewer

### ISSUE THREAD

**Issue by dinac** [daid/LegacyCura#134]
1.It will be really nice if there will be option in CURA  that shows in real time progress layer per layer in "GCode view"
This is great, you can go through layers ans seewhats happening, but you never know what layer is next

2.Another thing is maybe real time toolpath view ?
1. manual settings - override  in the real time - like temperature & speed ?

thanks !!

**Comment by rewolff** [comment]
Yes. It would be really nice. However, the whole system is modular. So the cura-core does rendering of the model and the resulting G-code. It calls external programs to slice and print the model.

For this to work, printrun will have to see the generated ";LAYER: xx" in the G-code  and for example update a "shared file" with the layer number.

**Comment by daid** [comment]
Rewolff, not everyone is printing with printrun. As Cura also has it's own printer interface (which has a few less features), which I assume this request is about.

I just added speed override settings to the Cura printer interface in the development version. I have not tested this with an actual print yet, so use at your own risk.

Adding the option to view the current location on the path while it is printing is too risky IMHO. I rather not risk slowing down the GCode sending by updating 3D preview windows. (Which is a real risk, as in RC2 the GCode sender delayed if you changed the model rotation)

**Comment by rewolff** [comment]
:-)
I'm still "new" in this end of the woods. I had misunderstood the "pinter interface in Cura" and "printrun" to be the same. Indeed by now I've found out that these are not the same. Sorry for the (possible) confusion. 
I'll see if I can test it tonight. 

Processes can be EITHER bandwidth or latency limited. 
If you make the delayed-acks too low, you'll be latency limited. Everytime you send a command, the 1ms USB clock needs to be waited for, the command needs to be sent. Then serialized for the AVR, intepreted, executed and then acked, serialized, usb-clock-wait again, and then operating system delay (if you don't set the proper flag in Linux up to 10ms). You'll be waiting for this whole loop for every command. Shave off 10 ms anywhere on the loop, and a 10k run will be 10k \* 10ms = 100 seconds faster.

But by setting the delayed acks parameter high enough, you'll become bandwidth limited. Add a whole second to that loop, and the run will become 1 second slower. 

If you carefully tune the delayed-acks to be only JUST high enough, any stealing of CPU time will immediately put you back in the delay limited regime and cause 
- longer print times
- visible defects on the print. 

If what I suspect is correct, the process doing the sending is also handling the gui of the print-module in cura. To prevent these "slowdowns" I would recommend the architecture that there is a background program without any gui that does the sending to the printer. It can for example memory map a communication file where it puts "int current_line;" at offset 0x00 and "float temperature;"  at 0x4. 
For dynamic adjusting of speed, we get the "float speedfactor = 1.0;" at offset 0x8.
Oh, And I' think I'd be very interested in a "last Z coordinate seen". 

You might think that keeping an eye on all these things takes too much time in the "send things to the printer" process. But these are just 4-byte writes to a shared memory region. That should take on the order of a few clocks of the multi-gigaherz host CPU. Way less than a single clock cycle on the arduino. 

I'm working on a library / system that makes these kinds of inter-process-communications-parameters easy. And the fact that the actual variables are  memory mapped is hidden from the user. It's not yet ready enough to be released to the public :-( . I'll see if I can whip something up for the cura print process.....

**Comment by daid** [comment]
The cura printing interface already runs as a separate process. There is also the python multiprocessing library: http://docs.python.org/library/multiprocessing.html which seems to handle this (I only found this a few days ago, still learning new things in python)

The reason why I separated the printing interface as separate process is because of the Python Interpreter Lock. Only once piece of python code can run at the same time. Which is fine for most of the GUI parts, in which the code runs quick and the GCode sender is available again before the serial buffer is exhausted. However, some parts, like GCode loading, and model transformations can hold the lock for a long time.

Keeping track of the Z height should be simple, and not require too much extra processing.

**Comment by rewolff** [comment]
OK. The speed adjustment thingy works. 

I've used it to see if a run would become better quality when running at 50mm/s instead of 100, as sliced. Then near the top the "min time per layer" kicked in, and it slowed down even more, while it could've continued to run at 50mm/sec IMHO. Anyway... "too complicated to get right" I'd say.... 
(NOW I wanted it to continue with "50 mm/sec" as the max speed. While someone else may want the behaviour I observed...)
